src/java.base/share/classes/jdk/internal/io/JdkConsoleImpl.java

```diff
@@ -56,6 +56,41 @@ public Reader reader() {
         return reader;
     }
 
+    @Override
+    public JdkConsole println(Object obj) {
+        pw.println(obj);
+        // automatic flushing covers println
+        return this;
+    }
+
+    @Override
+    public JdkConsole print(Object obj) {
+        pw.print(obj);
+        pw.flush(); // automatic flushing does not cover print
+        return this;
+    }
+
+    @Override
+    public String readln(String prompt) {
```

Comment:
this code can be simplified using an early return (and the body of the try/catch can be reduced so it is more clear which statement can cause the IOException)
```java
        synchronized (writeLock) {
            synchronized(readLock) {
                if (!prompt.isEmpty()) {
                    pw.print(prompt);
                    pw.flush(); // automatic flushing does not cover print
                }
               char[] array;
                try {
                    array = readline(false);
                } catch (IOException x) {
                    throw new IOError(x);
                }
                if (array != null) {
                    return new String(array);
                }
            }
        }
        return null;
```

Reply:
This method started as a copy of `readLine`. In its current form, it's very evident how `readln` differs from `readLine`, and I would leave it this way for now. If the feature is standardised, we could refactor both methods together, as you suggested.

If the `java.io.IO` part of that feature or the feature itself is withdrawn, then we could consider refactoring the existing `readLine`. Does it make sense?

Reply:
> I do not think the step to "standardise" a preview feature exists ? When a preview feature becomes a released feature, the code is very lightly edited, at least it this is my experience.

We may call it differently, but I think both you and I are referring to this part of [JEP 12](https://openjdk.org/jeps/12) (emphasis mine):

> Eventually, the JEP owner must decide the preview feature's fate. If the decision is to remove the preview feature, then the owner must file an issue in JBS to remove the feature in the next JDK feature release; no new JEP is needed. **On the other hand, if the decision is to finalize, then the owner must file a new JEP, noting refinements informed by developer feedback. The title of this JEP should be the feature's name, omitting the earlier suffix of (Preview) / (Second Preview), and without adding any new suffix such as (Standard) or (Final). This JEP will ultimately reach Targeted status for the next JDK feature release.**

> You can change both readln and readLine and if `java.io.IO` is removed, at least the code of readLine() will be

Sorry, Rémi, but no. As long as this feature is in preview, I'd optimise for easier removal (back out) of the feature rather than clean combined code.

Reply:
Okay